nexus/reconfigurator/planning/src/blueprint_builder/builder/datasets_editor.rs

```diff
@@ -0,0 +1,358 @@
+use nexus_types::deployment::BlueprintDatasetDisposition;
+use nexus_types::deployment::BlueprintDatasetsConfig;
+use nexus_types::deployment::SledResources;
+use nexus_types::deployment::ZpoolFilter;
+use omicron_common::api::external::ByteCount;
+use omicron_common::api::external::Generation;
+use omicron_common::disk::CompressionAlgorithm;
+use omicron_common::disk::DatasetKind;
+use omicron_common::disk::DatasetName;
+use omicron_common::disk::GzipLevel;
+use omicron_uuid_kinds::DatasetUuid;
+use omicron_uuid_kinds::SledUuid;
+use omicron_uuid_kinds::ZpoolUuid;
+use std::collections::btree_map::Entry;
+use std::collections::BTreeMap;
+use std::collections::BTreeSet;
+use std::net::SocketAddrV6;
+
+#[derive(Debug, thiserror::Error)]
+pub enum BlueprintDatasetsEditError {
+    #[error(
+        "internal inconsistency: multiple datasets with kind {kind:?} \
+         on zpool {zpool_id}: {id1}, {id2}"
+    )]
+    MultipleDatasetsOfKind {
+        zpool_id: ZpoolUuid,
+        kind: DatasetKind,
+        id1: DatasetUuid,
+        id2: DatasetUuid,
+    },
+}
+
+/// Helper for working with sets of datasets on each sled
+///
+/// Tracking the set of datasets is slightly non-trivial because we need to
+/// bump the per-sled generation number iff the datasets are changed.  So
+/// we need to keep track of whether we've changed the datasets relative
+/// to the parent blueprint.
+#[derive(Debug)]
+pub(super) struct BlueprintDatasetsEditor {
+    current: BTreeMap<SledUuid, BlueprintDatasetsConfig>,
+    changed: BTreeSet<SledUuid>,
+}
+
+impl BlueprintDatasetsEditor {
+    pub fn new(current: BTreeMap<SledUuid, BlueprintDatasetsConfig>) -> Self {
+        Self { current, changed: BTreeSet::new() }
+    }
+
+    /// Get a helper to edit the datasets of a specific sled.
+    ///
+    /// If any changes are made via the returned editor, the sled will be
+    /// recorded as needing a generation bump in its dataset config when the
+    /// editor is dropped.
+    pub fn sled_datasets_editor<'a>(
+        &'a mut self,
+        sled_id: SledUuid,
+        sled_resources: &SledResources,
+        rng: &'a mut PlannerRng,
+    ) -> Result<SledDatasetsEditor<'a>, BlueprintDatasetsEditError> {
+        let config = self
+            .current
+            .entry(sled_id)
+            .or_insert_with(empty_blueprint_datasets_config);
+
+        // Gather all dataset IDs known to the blueprint and to the database.
+        let blueprint_dataset_ids =
+            build_dataset_kind_id_map(config.datasets.values().map(
+                |dataset| (dataset.pool.id(), dataset.kind.clone(), dataset.id),
+            ))?;
+        let database_dataset_ids = build_dataset_kind_id_map(
+            sled_resources.all_datasets(ZpoolFilter::InService).flat_map(
+                |(&zpool_id, configs)| {
+                    configs.iter().map(move |config| {
+                        (zpool_id, config.name.dataset().clone(), config.id)
```

Comment:
I think there's a comment about it, but we should really change the name of the `dataset` method to be `dataset_kind`. Doesn't have to be (and probably shouldn't be) done in this PR, but it makes it look like `database_dataset_ids`  and `bluerprint_dataset_ids` are of different types at a glance.

Reply:
Agreed on "we should do it" and "not in this PR"  - filed #7115